ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/vm/VmTraceGenerator.java

```diff
@@ -0,0 +1,198 @@
+ */
+package org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.tracing.vm;
+
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.processor.TransactionTrace;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.tracing.Trace;
+import org.hyperledger.besu.ethereum.core.Gas;
+import org.hyperledger.besu.ethereum.debug.TraceFrame;
+import org.hyperledger.besu.ethereum.vm.Code;
+import org.hyperledger.besu.util.bytes.Bytes32;
+import org.hyperledger.besu.util.bytes.BytesValue;
+import org.hyperledger.besu.util.bytes.BytesValues;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+public class VmTraceGenerator {
+
+  /**
+   * Generate a stream of trace result objects.
+   *
+   * @param transactionTrace the transaction trace to use.
+   * @return a representation of generated traces.
+   */
+  public static Stream<Trace> generateTraceStream(final TransactionTrace transactionTrace) {
+    return Stream.of(generateTrace(transactionTrace));
+  }
+
+  /**
+   * Generate trace representation from the specified transaction trace.
+   *
+   * @param transactionTrace the transaction trace to use.
+   * @return a representation of the trace.
+   */
+  public static Trace generateTrace(final TransactionTrace transactionTrace) {
+    final VmTrace rootVmTrace = new VmTrace();
+    final Deque<VmTrace> parentTraces = new ArrayDeque<>();
+    parentTraces.add(rootVmTrace);
+    if (transactionTrace != null && !transactionTrace.getTraceFrames().isEmpty()) {
+      transactionTrace
+          .getTransaction()
+          .getInit()
+          .map(BytesValue::getHexString)
+          .ifPresent(rootVmTrace::setCode);
+      final AtomicInteger index = new AtomicInteger(0);
+      transactionTrace
+          .getTraceFrames()
+          .forEach(traceFrame -> addFrame(index, transactionTrace, traceFrame, parentTraces));
+    }
+    return rootVmTrace;
+  }
+
+  /**
+   * Add a trace frame to the VmTrace result object.
+   *
+   * @param index index of the current frame in the trace
+   * @param transactionTrace the transaction trace
+   * @param traceFrame the current trace frame
+   * @param parentTraces the queue of parent traces
+   */
+  private static void addFrame(
+      final AtomicInteger index,
+      final TransactionTrace transactionTrace,
+      final TraceFrame traceFrame,
+      final Deque<VmTrace> parentTraces) {
+    if ("STOP".equals(traceFrame.getOpcode())) {
+      return;
+    }
+
+    VmTrace newSubTrace;
+    VmTrace currentTrace = parentTraces.getLast();
+
+    // set smart contract code
+    currentTrace.setCode(traceFrame.getMaybeCode().orElse(new Code()).getBytes().getHexString());
+    final int nextFrameIndex = index.get() + 1;
+    // retrieve next frame if not last
+    final Optional<TraceFrame> maybeNextFrame =
```

Comment:
Instead of doing all of these comparisons here with `maybeNextFrame`, can we just capture more data in the `TraceFrame`'s?

Reply:
I would stay like that if you don't mind. I have reduced the number of comparisons with `maybeNextFrame `

Reply:
This is just a general suggestion to think about ways of making this more readable / pushing more functionality down into the core tracing utilities.  Not really a blocker.

Reply:
ok cool